library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/api/MutableResourcePack.java

```diff
@@ -0,0 +1,307 @@
+	 * @param id   the path of the resource
+	 * @param text the resource content
+	 * @see #putResource(ResourceType, Identifier, byte[])
+	 */
+	default void putText(@NotNull ResourceType type, @NotNull Identifier id, @NotNull String text) {
+		this.putResource(type, id, text.getBytes(StandardCharsets.UTF_8));
+	}
+
+	/**
+	 * Puts a text resource into the resource pack's root.
+	 *
+	 * @param fileName     the name of the file
+	 * @param textSupplier the supplier of the resource content
+	 * @apiNote the supplier is {@link com.google.common.base.Suppliers#memoize(com.google.common.base.Supplier) memoized}
+	 * @see #putResource(String, Supplier)
+	 */
+	default void putText(@NotNull String fileName, @NotNull Supplier<@NotNull String> textSupplier) {
+		this.putResource(fileName, () -> textSupplier.get().getBytes(StandardCharsets.UTF_8));
+	}
+
+	/**
+	 * Puts a text resource into the resource pack for the given side and path.
+	 *
+	 * @param type         the resource type
+	 * @param id           the path of the resource
+	 * @param textSupplier the supplier of the resource content
+	 * @apiNote the supplier is {@link com.google.common.base.Suppliers#memoize(com.google.common.base.Supplier) memoized}
+	 * @see #putResource(ResourceType, Identifier, Supplier)
+	 */
+	default void putText(@NotNull ResourceType type, @NotNull Identifier id, @NotNull Supplier<@NotNull String> textSupplier) {
+		this.putResource(type, id, () -> textSupplier.get().getBytes(StandardCharsets.UTF_8));
+	}
+
+	/**
+	 * Puts a resource into the resource pack's root asynchronously.
+	 *
+	 * @param fileName    the name of the file
+	 * @param textFactory the factory of the resource content
+	 * @return the future
+	 * @see #putResourceAsync(String, Function)
+	 */
+	default @NotNull Future<byte[]> putTextAsync(@NotNull String fileName, @NotNull Function<@NotNull String, @NotNull String> textFactory) {
+		return this.putResourceAsync(fileName, textFactory.andThen(text -> text.getBytes(StandardCharsets.UTF_8)));
+	}
+
+	/**
+	 * Puts a text resource into the resource pack for the given side and path asynchronously.
+	 *
+	 * @param type        the resource type
+	 * @param id          the path of the resource
+	 * @param textFactory the factory of the resource content
+	 * @return the future
+	 * @see #putResourceAsync(ResourceType, Identifier, Function)
+	 */
+	default @NotNull Future<byte[]> putTextAsync(@NotNull ResourceType type, @NotNull Identifier id,
+			@NotNull Function<@NotNull Identifier, @NotNull String> textFactory) {
+		return this.putResourceAsync(type, id, textFactory.andThen(text -> text.getBytes(StandardCharsets.UTF_8)));
+	}
+
+	/**
+	 * Puts an image resource into the resource pack's root.
+	 * <p>
+	 * <b>Note:</b> this method is only available on the client.
+	 *
+	 * @param fileName the name of the file
+	 * @param image    the resource content
+	 * @see #putResource(String, byte[])
+	 */
+	@Environment(EnvType.CLIENT)
```

Comment:
It seems weird to me that all the methods here are made to be side-agnostic except these ones. I'm not really sure why these types were included in particular. `NativeImage` doesn't really seem like a useful format to be working with for dynamic images either. `ImageIO` includes more tools to edit images at least.

Reply:
I personally am using NativeImage for anything image-related in MC.
And it's a client-only class so those methods must be client-only.

I usually avoid touching anything in the java.awt package due to the famous macOs issues.
It shouldn't happen with ImageIO though, so idk.
I've always found the BufferedImages more annoying to work with compared to NativeImage, at least one the format is easy.

Reply:
Fair enough, I guess that would be personal preference, and most of the actually useful image editing is actually in `java.awt` so it might not be the best of bets. I really like how most of the stuff in here is side-agnostic (for PolyMc hackery reasons) so I was just curious about this one case. I just hope people won't start registering their resource packs inside the client mod initializer but I know they'll probably do that anyway.